docs/dev/shims.md

```diff
@@ -48,8 +48,8 @@ inject an intermediate trait e.g. `com.nvidia.spark.rapids.shims.ShimExpression`
 has a varying source code depending on the Spark version we compile against to overcome this
```

Comment:
This document needs extensive updates to discuss shimplify (maybe it warrants its own separate document referenced here instead?) which covers all the options, details on what it does, the overview of what happens during the build process, how debugging shim code in the IDE changes (if that's necessary to cover), etc.

I expect this document to cover everything detailed in the PR description and much more.  Ideally everything one would want to know about shimplify and how the shims work should be discoverable from here.